py/desitarget/targets.py

```diff
@@ -387,6 +390,90 @@ def initial_priority_numobs(targets, scnd=False,
     return outpriority, outnumobs
 
 
+def calc_numobs_more(targets, zcat, obscon):
+    """
+    Calculate target NUMOBS_MORE from masks, observation/redshift status.
+
+    Parameters
+    ----------
+    targets : :class:`~numpy.ndarray`
+        numpy structured array or astropy Table of targets. Must include
+        the columns `DESI_TARGET`, `BGS_TARGET`, `MWS_TARGET`
+        (or their SV/cmx equivalents).
+    zcat : :class:`~numpy.ndarray`
+        numpy structured array or Table of redshift information. Must
+        include 'Z', `ZWARN`, `NUMOBS` and be the same length as
+        `targets`. May also contain `NUMOBS_MORE` if this isn't the
+        first time through MTL and `NUMOBS > 0`.
+    obscon : :class:`str`
+        A combination of strings that are in the desitarget bitmask yaml
+        file (specifically in `desitarget.targetmask.obsconditions`), e.g.
+        "DARK|GRAY". Governs the behavior of how priorities are set based
+        on "obsconditions" in the desitarget bitmask yaml file.
+
+    Returns
+    -------
+    :class:`~numpy.array`
+        Integer array of number of additional observations (NUMOBS_MORE).
+
+    Notes
+    -----
+        - Will automatically detect if the passed targets are main
+          survey, commissioning or SV and behave accordingly.
+        - Most targets are updated to NUMOBS_MORE = NUMOBS_INIT-NUMOBS.
+          Special cases include BGS targets which always get NUMOBS_MORE
+          of 1 in bright time and QSO "tracer" targets which always get
+          NUMOBS_MORE=0 in dark time.
+    """
+    # ADM check the input arrays are the same length.
+    assert len(targets) == len(zcat)
+
+    # ADM determine whether the input targets are main survey, cmx or SV.
+    colnames, masks, survey = main_cmx_or_sv(targets, scnd=True)
+    # ADM the target bits/names should be shared between main survey and SV.
+    if survey != 'cmx':
+        desi_target, bgs_target, mws_target, scnd_target = colnames
+        desi_mask, bgs_mask, mws_mask, scnd_mask = masks
+    else:
+        cmx_mask = masks[0]
+
+    # ADM Default is 0 , i.e. no more observations.
+    numobs_more = np.zeros(len(targets), dtype='i8')
+
+    # ADM main case, just decrement by NUMOBS.
+    numobs_more = np.maximum(0, targets['NUMOBS_INIT'] - zcat['NUMOBS'])
```

Comment:
We have to keep in mind that this logic will give zero `numobs_more` to all quasars if we start QSOs with NUMOBS_INIT=1. (https://github.com/desihub/desitarget/issues/616).

Reply:
Agreed. If we do change to `NUMOBS_INIT=1` for QSOs, we will need more special logic in the new `targets.calc_numobs_more()` function (which we would have needed anyway, previously, in the `mtl.make_mtl()` function).